server/auto_updater/dataset_updater.py

```diff
@@ -0,0 +1,301 @@
+        current_pmc_file - The listing of papers in PMCOA
+        word_model - The word2vec model to generate paper embeddings
+        temp_dir_filename - a temp file contains newly added papers
+        temp_embed_filename - a file containing new paper embeddingss
+        temp_token_counter - a file containing new tokens from processed papers
+    """
+
+    # Grab Current List
+    current_pmc_list_df = pd.read_csv(current_pmc_file, sep="\t")
+    current_pmc_set = set(current_pmc_list_df.file_path.tolist())
+
+    # Grab file list on ftp server
+    pmc_open_access_url = "ftp://ftp.ncbi.nlm.nih.gov/pub/pmc/oa_bulk/"
+    response = request.urlopen(f"{pmc_open_access_url}")
+    files = response.read().decode("utf-8").splitlines()
+    tar_files = [f.split(" ")[-1] for f in files]
+
+    with open(temp_dir_filename, "w") as dir_file, open(
+        temp_embed_filename, "w"
+    ) as embed_file, open(temp_token_counter, "w") as word_file:
+        dir_writer = csv.DictWriter(
+            dir_file, delimiter="\t", fieldnames=["tarfile", "file_path"]
+        )
+        dir_writer.writeheader()
+
+        embed_writer = csv.DictWriter(
+            embed_file,
+            delimiter="\t",
+            fieldnames=["journal", "document"] + [f"feat_{idx}" for idx in range(300)],
+        )
+        embed_writer.writeheader()
+
+        count_writer = csv.DictWriter(
+            word_file, delimiter="\t", fieldnames=["document", "lemma", "count"]
+        )
+        count_writer.writeheader()
+
+        # Cycle through each tar file on the server
+        for tar_file in tqdm.tqdm(tar_files):
+
+            # If not xml files skip
+            if all(suffix != ".xml" for suffix in Path(tar_file).suffixes):
+                continue
+
+            # If temp file skip
+            if Path(tar_file).suffix == ".tmp":
+                continue
+
+            # Grab the file from the tarfile
+            print(f"Requesting {pmc_open_access_url}{tar_file}....")
+            requested_file_stream = request.urlopen(f"{pmc_open_access_url}{tar_file}")
+            open_stream = tarfile.open(fileobj=requested_file_stream, mode="r|gz")
+
+            # Cycle through paper members
+            # Grab new papers and get the document vectors
+            while True:
+                try:
+
+                    pmc_paper = open_stream.next()
+
+                    if pmc_paper is None:
+                        break
+
+                    if pmc_paper.isdir():
+                        continue
+
+                    if pmc_paper.name in current_pmc_set:
+                        continue
+
+                    new_paper = open_stream.extractfile(pmc_paper)
+                    doc_vector, word_counter = generate_vector_counts(
+                        word_model, new_paper, "//abstract/sec/*|//body/sec/*"
+                    )
+
+                    if word_counter is None:
+                        continue
+
+                    dir_writer.writerow(
```

Comment:
I think this statement (lines 137-139) should be moved right after line 132, because even if `word_counter` is `None` in the current paper, the paper should be still recorded in `new_papers_added_dir.tsv`; otherwise when the updater is executed sometime in the future, this paper will be parsed again and get skipped again, which is a waste of computational resource.